nacos ipv6 register

## Inacos ipv6 register failed



1.  nacos register with ipv4 is normal
2.  However,  nacos register with ipv6 connect failed.
3.  and the remote log such as:

2022-10-08 15:57:03.305 INFO [main :c.a.n.c.r.client] [48e91b39-50fc-4fab-925d-4267f75b7cb7] Try to connect to server on start up, server: {serverIp = '[fe80', server main port = 8848}
2022-10-08 15:57:03.306 ERROR [main :c.a.n.c.r.c.g.GrpcClient] [48e91b39-50fc-4fab-925d-4267f75b7cb7]Fail to connect to server!,error={}
java.lang.IllegalArgumentException: Invalid host or port: [fe80 9848
	at com.alibaba.nacos.shaded.io.grpc.internal.GrpcUtil.authorityFromHostAndPort(GrpcUtil.java:488) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.shaded.io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder.<init>(NettyChannelBuilder.java:133) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.shaded.io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder.forAddress(NettyChannelBuilder.java:119) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.shaded.io.grpc.netty.shaded.io.grpc.netty.NettyChannelProvider.builderForAddress(NettyChannelProvider.java:37) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.shaded.io.grpc.netty.shaded.io.grpc.netty.NettyChannelProvider.builderForAddress(NettyChannelProvider.java:23) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.shaded.io.grpc.ManagedChannelBuilder.forAddress(ManagedChannelBuilder.java:39) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.common.remote.client.grpc.GrpcClient.createNewChannelStub(GrpcClient.java:143) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.common.remote.client.grpc.GrpcClient.connectToServer(GrpcClient.java:299) ~[nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.common.remote.client.RpcClient.start(RpcClient.java:389) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.client.naming.remote.gprc.NamingGrpcClientProxy.start(NamingGrpcClientProxy.java:95) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.client.naming.remote.gprc.NamingGrpcClientProxy.<init>(NamingGrpcClientProxy.java:88) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.client.naming.remote.NamingClientProxyDelegate.<init>(NamingClientProxyDelegate.java:76) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.client.naming.NacosNamingService.init(NacosNamingService.java:95) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.client.naming.NacosNamingService.<init>(NacosNamingService.java:81) [nacos-client-2.1.0.jar:na]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) [na:1.8.0_171]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source) [na:1.8.0_171]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source) [na:1.8.0_171]
	at java.lang.reflect.Constructor.newInstance(Unknown Source) [na:1.8.0_171]
	at com.alibaba.nacos.api.naming.NamingFactory.createNamingService(NamingFactory.java:59) [nacos-client-2.1.0.jar:na]
	at com.alibaba.nacos.api.NacosFactory.createNamingService(NacosFactory.java:77) [nacos-client-2.1.0.jar:na]
	at com.alibaba.cloud.nacos.NacosServiceManager.createNewNamingService(NacosServiceManager.java:99) [spring-cloud-starter-alibaba-nacos-discovery-2.2.9.RELEASE.jar:2.2.9.RELEASE]
	at com.alibaba.cloud.nacos.NacosServiceManager.buildNamingService(NacosServiceManager.java:90) [spring-cloud-starter-alibaba-nacos-discovery-2.2.9.RELEASE.jar:2.2.9.RELEASE]
	at com.alibaba.cloud.nacos.NacosServiceManager.getNamingService(NacosServiceManager.java:46) [spring-cloud-starter-alibaba-nacos-discovery-2.2.9.RELEASE.jar:2.2.9.RELEASE]
	at com.alibaba.cloud.nacos.discovery.NacosWatch.start(NacosWatch.java:131) [spring-cloud-starter-alibaba-nacos-discovery-2.2.9.RELEASE.jar:2.2.9.RELEASE]
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:178) [spring-context-5.3.15.jar:5.3.15]
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:54) [spring-context-5.3.15.jar:5.3.15]
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:356) [spring-context-5.3.15.jar:5.3.15]
	at java.lang.Iterable.forEach(Unknown Source) ~[na:1.8.0_171]
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:155) [spring-context-5.3.15.jar:5.3.15]
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:123) [spring-context-5.3.15.jar:5.3.15]
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:935) ~[spring-context-5.3.15.jar:5.3.15]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:586) ~[spring-context-5.3.15.jar:5.3.15]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.6.3.jar:2.6.3]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:732) ~[spring-boot-2.6.3.jar:2.6.3]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:414) ~[spring-boot-2.6.3.jar:2.6.3]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:302) ~[spring-boot-2.6.3.jar:2.6.3]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1303) ~[spring-boot-2.6.3.jar:2.6.3]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1292) ~[spring-boot-2.6.3.jar:2.6.3]
	at com.cmcc.emdp.EmdpFtpApplication.main(EmdpFtpApplication.java:24) ~[classes/:na]
Caused by: java.net.URISyntaxException: Expected closing bracket for IPv6 address at index 12: //[fe80:9848
	at java.net.URI$Parser.fail(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI$Parser.failExpecting(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI$Parser.parseServer(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI$Parser.parseAuthority(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI$Parser.parseHierarchical(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI$Parser.parse(Unknown Source) ~[na:1.8.0_171]
	at java.net.URI.<init>(Unknown Source) ~[na:1.8.0_171]
	at com.alibaba.nacos.shaded.io.grpc.internal.GrpcUtil.authorityFromHostAndPort(GrpcUtil.java:486) ~[nacos-client-2.1.0.jar:na]
	... 38 common frames omitted

--- a/common/src/main/java/com/alibaba/nacos/common/remote/client/RpcClient.java
+++ b/common/src/main/java/com/alibaba/nacos/common/remote/client/RpcClient.java
@@ -33,6 +33,7 @@ import com.alibaba.nacos.common.lifecycle.Closeable;
 import com.alibaba.nacos.common.remote.ConnectionType;
 import com.alibaba.nacos.common.remote.PayloadRegistry;
 import com.alibaba.nacos.common.utils.CollectionUtils;
+import com.alibaba.nacos.common.utils.InternetAddressUtil;
 import com.alibaba.nacos.common.utils.LoggerUtils;
 import com.alibaba.nacos.common.utils.NumberUtils;
 import com.alibaba.nacos.common.utils.StringUtils;
@@ -895,8 +896,7 @@ public abstract class RpcClient implements Closeable {
         if (matcher.find()) {
             serverAddress = matcher.group(1);
         }
-        
-        String[] ipPortTuple = serverAddress.split(Constants.COLON, 2);
+        String[] ipPortTuple = InternetAddressUtil.splitIPPortStr(serverAddress);
         int defaultPort = Integer.parseInt(System.getProperty("nacos.server.port", "8848"));
         String serverPort = CollectionUtils.getOrDefault(ipPortTuple, 1, Integer.toString(defaultPort));
         

--- a/common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java
+++ b/common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java
@@ -122,6 +122,8 @@ public abstract class GrpcClient extends RpcClient {
     }
     
     protected ThreadPoolExecutor createGrpcExecutor(String serverIp) {
+        // Thread name will use String.format, ipv6 maybe contain special word %, so handle it first.
+        serverIp = serverIp.replaceAll("%", "-");
         ThreadPoolExecutor grpcExecutor = new ThreadPoolExecutor(clientConfig.threadPoolCoreSize(),
                 clientConfig.threadPoolMaxSize(), clientConfig.threadPoolKeepAlive(), TimeUnit.MILLISECONDS,
                 new LinkedBlockingQueue<>(clientConfig.threadPoolQueueSize()),

--- a/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java
+++ b/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java
@@ -21,6 +21,7 @@ import com.alibaba.nacos.api.remote.RequestCallBack;
 import com.alibaba.nacos.api.remote.request.Request;
 import com.alibaba.nacos.api.remote.response.ErrorResponse;
 import com.alibaba.nacos.common.remote.ConnectionType;
+import com.alibaba.nacos.common.remote.client.grpc.DefaultGrpcClientConfig;
 import com.alibaba.nacos.common.remote.client.grpc.GrpcConnection;
 import org.junit.After;
 import org.junit.Assert;
@@ -41,6 +42,7 @@ import java.util.Map;
 import java.util.Queue;
 import java.util.Random;
 
+import static org.junit.Assert.assertEquals;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.atLeastOnce;
@@ -171,11 +173,11 @@ public class RpcClientTest {
     public void testInitServerListFactory() {
         rpcClient.rpcClientStatus.set(RpcClientStatus.WAIT_INIT);
         rpcClient.serverListFactory(serverListFactory);
-        Assert.assertEquals(RpcClientStatus.INITIALIZED, rpcClient.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.INITIALIZED, rpcClient.rpcClientStatus.get());
         
         rpcClient.rpcClientStatus.set(RpcClientStatus.INITIALIZED);
         rpcClient.serverListFactory(serverListFactory);
-        Assert.assertEquals(RpcClientStatus.INITIALIZED, rpcClient.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.INITIALIZED, rpcClient.rpcClientStatus.get());
         
         RpcClient client1 = new RpcClient(new RpcClientConfig() {
             @Override
@@ -228,7 +230,7 @@ public class RpcClientTest {
                 return null;
             }
         };
-        Assert.assertEquals(RpcClientStatus.INITIALIZED, client1.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.INITIALIZED, client1.rpcClientStatus.get());
         
         RpcClient client2 = new RpcClient(rpcClientConfig, serverListFactory) {
             @Override
@@ -246,21 +248,21 @@ public class RpcClientTest {
                 return null;
             }
         };
-        Assert.assertEquals(RpcClientStatus.INITIALIZED, client2.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.INITIALIZED, client2.rpcClientStatus.get());
     }
     
     @Test
     public void testLabels() {
         when(rpcClientConfig.labels()).thenReturn(Collections.singletonMap("labelKey1", "labelValue1"));
         Map.Entry<String, String> element = rpcClient.getLabels().entrySet().iterator().next();
-        Assert.assertEquals("labelKey1", element.getKey());
-        Assert.assertEquals("labelValue1", element.getValue());
+        assertEquals("labelKey1", element.getKey());
+        assertEquals("labelValue1", element.getValue());
         
         // accumulate labels
         Map<String, String> map = new HashMap<>();
         map.put("labelKey2", "labelValue2");
         when(rpcClientConfig.labels()).thenReturn(map);
-        Assert.assertEquals(1, rpcClient.getLabels().size());
+        assertEquals(1, rpcClient.getLabels().size());
     }
     
     @Test
@@ -271,7 +273,7 @@ public class RpcClientTest {
         rpcClient.onServerListChange();
         
         int afterSize = ((Queue<?>) reconnectionSignalField.get(rpcClient)).size();
-        Assert.assertEquals(beforeSize, afterSize);
+        assertEquals(beforeSize, afterSize);
     }
     
     @Test
@@ -282,7 +284,7 @@ public class RpcClientTest {
         rpcClient.onServerListChange();
         
         int afterSize = ((Queue<?>) reconnectionSignalField.get(rpcClient)).size();
-        Assert.assertEquals(beforeSize, afterSize);
+        assertEquals(beforeSize, afterSize);
     }
     
     @Test
@@ -292,11 +294,11 @@ public class RpcClientTest {
         rpcClient.serverListFactory(serverListFactory);
         rpcClient.currentConnection = new GrpcConnection(new RpcClient.ServerInfo("10.10.10.10", 8848), null);
         doReturn(Collections.singletonList("")).when(serverListFactory).getServerList();
-        
+        when(serverListFactory.getServerList()).thenReturn(Collections.singletonList("127.0.0.1"));
         rpcClient.onServerListChange();
         
         int afterSize = ((Queue<?>) reconnectionSignalField.get(rpcClient)).size();
-        Assert.assertEquals(beforeSize + 1, afterSize);
+        assertEquals(beforeSize + 1, afterSize);
     }
     
     @Test
@@ -305,35 +307,35 @@ public class RpcClientTest {
         rpcClient.serverListFactory(serverListFactory);
         rpcClient.currentConnection = new GrpcConnection(new RpcClient.ServerInfo("10.10.10.10", 8848), null);
         doReturn(Collections.singletonList("http://10.10.10.10:8848")).when(serverListFactory).getServerList();
-        
         rpcClient.onServerListChange();
         
         int afterSize = ((Queue<?>) reconnectionSignalField.get(rpcClient)).size();
-        Assert.assertEquals(beforeSize, afterSize);
+        assertEquals(beforeSize, afterSize);
     }
     
     @Test
     public void testResolveServerInfo1() throws InvocationTargetException, IllegalAccessException {
-        Assert.assertEquals(":8848",
-                ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848",
+        assertEquals("10.10.10.10:8848",
                 ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "10.10.10.10::8848")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848",
+        assertEquals("10.10.10.10:8848",
                 ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "10.10.10.10:8848")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848", ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient,
-                "http://10.10.10.10:8848")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848", ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient,
-                "http://10.10.10.10::8848")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848",
+        assertEquals("10.10.10.10:8848",
+                ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "http://10.10.10.10:8848"))
+                        .getAddress());
+        assertEquals("10.10.10.10:8848",
+                ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "http://10.10.10.10::8848"))
+                        .getAddress());
+        assertEquals("10.10.10.10:8848",
                 ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "http://10.10.10.10")).getAddress());
-        Assert.assertEquals("10.10.10.10:8848", ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient,
-                "https://10.10.10.10::8848")).getAddress());
+        assertEquals("10.10.10.10:8848",
+                ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "https://10.10.10.10::8848"))
+                        .getAddress());
     }
     
     @Test
     public void testResolveServerInfo2() throws InvocationTargetException, IllegalAccessException {
         System.setProperty("nacos.server.port", "4424");
-        Assert.assertEquals("10.10.10.10:4424",
+        assertEquals("10.10.10.10:4424",
                 ((RpcClient.ServerInfo) resolveServerInfoMethod.invoke(rpcClient, "http://10.10.10.10")).getAddress());
     }
     
@@ -377,7 +379,7 @@ public class RpcClientTest {
             exception = e;
         }
         
-        Assert.assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
         verify(rpcClient).switchServerAsync();
         Assert.assertNotNull(exception);
     }
@@ -409,7 +411,7 @@ public class RpcClientTest {
         verify(connection, atLeastOnce()).asyncRequest(any(), any());
         verify(rpcClient).switchServerAsyncOnRequestFail();
         Assert.assertNotNull(exception);
-        Assert.assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
     }
     
     @Test(expected = NacosException.class)
@@ -438,7 +440,7 @@ public class RpcClientTest {
         verify(connection, times(3)).requestFuture(any());
         verify(rpcClient).switchServerAsyncOnRequestFail();
         Assert.assertNotNull(exception);
-        Assert.assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
+        assertEquals(RpcClientStatus.UNHEALTHY, rpcClient.rpcClientStatus.get());
     }
     
     @Test
@@ -512,5 +514,125 @@ public class RpcClientTest {
             e.printStackTrace();
         }
         verify(connection, times(retry + 1)).request(any(), anyLong());
+    }
+    
+    @Test
+    public void testNextRpcServerForIpv4WithPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("127.0.0.1:7777");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("127.0.0.1:7777", actual.getAddress());
+        assertEquals("127.0.0.1", actual.getServerIp());
+        assertEquals(7777, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForIpv4WithoutPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("127.0.0.1");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("127.0.0.1:8848", actual.getAddress());
+        assertEquals("127.0.0.1", actual.getServerIp());
+        assertEquals(8848, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForIpv6WithPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("[fe80::35ba:6827:c5ff:d161%11]:7777");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("[fe80::35ba:6827:c5ff:d161%11]:7777", actual.getAddress());
+        assertEquals("[fe80::35ba:6827:c5ff:d161%11]", actual.getServerIp());
+        assertEquals(7777, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForIpv6WithoutPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("[fe80::35ba:6827:c5ff:d161%11]");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("[fe80::35ba:6827:c5ff:d161%11]:8848", actual.getAddress());
+        assertEquals("[fe80::35ba:6827:c5ff:d161%11]", actual.getServerIp());
+        assertEquals(8848, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForDomainWithPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("nacos.io:7777");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("nacos.io:7777", actual.getAddress());
+        assertEquals("nacos.io", actual.getServerIp());
+        assertEquals(7777, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForDomainWithoutPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("nacos.io");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("nacos.io:8848", actual.getAddress());
+        assertEquals("nacos.io", actual.getServerIp());
+        assertEquals(8848, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForLocalhostWithPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("localhost:7777");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("localhost:7777", actual.getAddress());
+        assertEquals("localhost", actual.getServerIp());
+        assertEquals(7777, actual.getServerPort());
+    }
+    
+    @Test
+    public void testNextRpcServerForLocalhostWithoutPort() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("localhost");
+        RpcClient.ServerInfo actual = rpcClient.nextRpcServer();
+        assertEquals("localhost:8848", actual.getAddress());
+        assertEquals("localhost", actual.getServerIp());
+        assertEquals(8848, actual.getServerPort());
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void testNextRpcServerForEmpty() {
+        RpcClient rpcClient = buildTestNextRpcServerClient();
+        rpcClient.serverListFactory(serverListFactory);
+        when(serverListFactory.genNextServer()).thenReturn("");
+        rpcClient.nextRpcServer();
+    }
+    
+    private RpcClient buildTestNextRpcServerClient() {
+        return new RpcClient(DefaultGrpcClientConfig.newBuilder().build()) {
+            @Override
+            public ConnectionType getConnectionType() {
+                return null;
+            }
+            
+            @Override
+            public int rpcPortOffset() {
+                return 0;
+            }
+            
+            @Override
+            public Connection connectToServer(ServerInfo serverInfo) throws Exception {
+                return null;
+            }
+            
+            @Override
+            public ServerInfo nextRpcServer() {
+                return super.nextRpcServer();
+            }
+        };
     }
 }

